### PR TITLE
Promoting EndpointSlices to beta

### PR DESCRIPTION
**Pull request by robscott**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
EndpointSlices provide significant scalability improvements as an alternative to Endpoints. The alpha release has proved itself well through a wide variety of scale testing, including tests up to 50k endpoints in 4k node clusters. Graduating EndpointSlices to beta will ensure that network endpoints are scalable and performant by default in Kubernetes.

**Special notes for your reviewer**:
The KEP outlines the graduation criteria for a beta release of EndpointSlices in 1.17. Presently all requirements are merged or have PRs ready for final review/approval. I'll keep this updated with the current status of each requirement:
- [x] Kube-proxy switch to consume EndpointSlice API. (Already done in Alpha)
- [x] Verify performance/scalability via testing. (Scale tested to 50k endpoints in 4k node cluster)
- [x] Get performance fixes identified in scale testing merged. (https://github.com/kubernetes/kubernetes/pull/83035, https://github.com/kubernetes/kubernetes/pull/83206, https://github.com/kubernetes/kubernetes/pull/83208)
- [x] Add FQDN addressType. (https://github.com/kubernetes/kubernetes/pull/84091)
- [x] Add support for optional appProtocol field on EndpointPort. (https://github.com/kubernetes/kubernetes/pull/83815)
- [x] Add support for endpointslice.kubernetes.io/managed-by label. (https://github.com/kubernetes/kubernetes/pull/83965)
- [x] Implement e2e tests that ensure both Endpoints and EndpointSlices are tested. (https://github.com/kubernetes/kubernetes/pull/84208)
- [ ] Implement dual-stack EndpointSlice support in kube-proxy. (https://github.com/kubernetes/kubernetes/pull/85246)

**Does this PR introduce a user-facing change?**:
```release-note
EndpointSlices are now beta and enabled by default for better Network Endpoint performance at scale.
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:
- Enhancement Issue: https://github.com/kubernetes/enhancements/issues/752
- KEP: https://github.com/kubernetes/enhancements/blob/master/keps/sig-network/20190603-EndpointSlice-API.md

/sig network api-machinery
/priority important-soon
/milestone v1.17

### CODE REVIEW

**Comment by robscott** [comment]
Since https://github.com/kubernetes/kubernetes/pull/84971 has taken longer than expected to merge in, I'm including the commit in this PR. It required a lot of changes to support the updates to AddressType in that PR and with code freeze coming up I wanted to ensure this had enough time for a proper review.

**Comment by robscott** [comment]
@thockin and @freehan do you have time to take a look at this PR?

/assign @thockin @freehan

**Comment by thockin** [comment]
Do we wait for the other PRs to merge before going with this?

File: pkg/proxy/winkernel/proxier_test.go [hunk 1]
```diff
@@ -20,7 +20,7 @@ package winkernel
 
 import (
 	"k8s.io/api/core/v1"
-	discovery "k8s.io/api/discovery/v1alpha1"
```
Comment (thockin) [inline pkg/proxy/winkernel/proxier_test.go]: At what stage do we add EPSlice support to Windows proxiers?

File: pkg/registry/discovery/rest/storage_discovery.go [hunk 2]
```diff
@@ -44,6 +45,14 @@ func (p StorageProvider) NewRESTStorage(apiResourceConfigSource serverstorage.AP
 		apiGroupInfo.VersionedResourcesStorageMap[discoveryv1alpha1.SchemeGroupVersion.Version] = storageMap
 	}
 
+	// if apiResourceConfigSource.VersionEnabled(discoveryv1beta1.SchemeGroupVersion) {
```
Comment (thockin) [inline pkg/registry/discovery/rest/storage_discovery.go]: what's with these comments?

Comment (robscott) [reply pkg/registry/discovery/rest/storage_discovery.go]: No idea why/when I did that, fixed. Thanks for catching this!

**Comment by robscott** [comment]
@thockin I don't think the order matters too much between this and the related proxy PR, I'd say whichever is ready first can go in first and then the other can update based on that. I'm going to be doing the final EndpointSlice + dual stack work on kube-proxy as well, so either way, it'll be me that has to deal with whichever one gets in first.

File: pkg/proxy/winkernel/proxier_test.go (hunk 1 above)
Comment (robscott) [reply pkg/proxy/winkernel/proxier_test.go]: Hopefully soon, maybe 1.18. I'll try to reach out to some of the people who are more familiar with that side of things to see if we can work together to get EndpointSlice integration in place.

**Review by thockin — COMMENTED** [review]
Thanks!

/lgtm
/approve

File: cmd/kube-controller-manager/app/discovery.go [hunk 3]
```diff
@@ -23,22 +23,15 @@ package app
 import (
 	"net/http"
 
-	discoveryv1alpha1 "k8s.io/api/discovery/v1alpha1"
-	"k8s.io/klog"
 	endpointslicecontroller "k8s.io/kubernetes/pkg/controller/endpointslice"
 )
 
 func startEndpointSliceController(ctx ControllerContext) (http.Handler, bool, error) {
-	if !ctx.AvailableResources[discoveryv1alpha1.SchemeGroupVersion.WithResource("endpointslices")] {
```
Comment (liggitt) [inline cmd/kube-controller-manager/app/discovery.go]: still need to check if the beta API is available... beta APIs can be turned off and components should still run

Comment (liggitt) [reply cmd/kube-controller-manager/app/discovery.go]: and check if the EndpointSlice feature is enabled before running

**Pull request merged**